File: bindings/typescript/settle/src/ddl.ts
import type { ColumnType } from './column'
import type { StateFieldDef } from './settle'

// ─── Duration parsing ────────────────────────────────────────────

const DURATION_UNITS: Record<string, number> = {
  second: 1,
  seconds: 1,
  sec: 1,
  s: 1,
  minute: 60,
  minutes: 60,
  min: 60,
  m: 60,
  hour: 3600,
  hours: 3600,
  hr: 3600,
  h: 3600,
  day: 86400,
  days: 86400,
  d: 86400,
}

export function parseDuration(s: string): number {
  const match = s.trim().match(/^(\d+)\s*(\w+)$/)
  if (!match) {
    throw new Error(`invalid duration: '${s}'`)
  }

  const n = parseInt(match[1], 10)
  const unit = match[2].toLowerCase()
  const mult = DURATION_UNITS[unit]
  if (!mult) {
    throw new Error(`unknown duration unit: '${unit}'`)
  }

  return n * mult
}

// ─── SQL escaping ───────────────────────────────────────────────

function sqlEscape(s: string): string {
  return s.replace(/'/g, "''")
}

// ─── Interval helper ─────────────────────────────────────────────

export interface IntervalExpr {
  _type: 'interval'
  column: string
  seconds: number
  alias?: string
  as(alias: string): IntervalExpr
}

function makeInterval(column: string, seconds: number, alias?: string): IntervalExpr {
  return {
    _type: 'interval',
    column,
    seconds,
    alias,
    as(a: string): IntervalExpr {
      return makeInterval(column, seconds, a)
    },
  }
}

export function interval(column: string, duration: string): IntervalExpr {
  return makeInterval(column, parseDuration(duration))
}

// ─── Aggregation types ───────────────────────────────────────────

export interface AggExpr {
  _type: 'agg'
  func: string
  column: string | null
}

export interface KeyRef {
  _type: 'key'
  column: string
}

export interface AggProxy<TSource = any> {
  key: Record<string, KeyRef>
  sum(column: string & keyof TSource): AggExpr
  count(): AggExpr
  first(column: string & keyof TSource): AggExpr
  last(column: string & keyof TSource): AggExpr
  min(column: string & keyof TSource): AggExpr
  max(column: string & keyof TSource): AggExpr
  avg(column: string & keyof TSource): AggExpr
}

export type GroupByItem = string | IntervalExpr

// ─── Options types ───────────────────────────────────────────────

export type ReducerCtx<TState, TEmit> = Readonly<TState> & {
  update(newState: TState): void
  emit(row: TEmit): void
}

export interface ReducerOptions<TState, TRow, TEmit> {
  groupBy: (string & keyof TRow) | (string & keyof TRow)[]
  initialState: TState
  reduce: (state: ReducerCtx<TState, TEmit>, row: TRow) => void
}

export interface SlidingWindowOptions<TSource = any> {
  /** Window duration, e.g. "1 hour", "30 minutes", "86400 seconds" */
  interval: string
  /** Column containing row timestamps (DateTime in milliseconds).
   *  Must be a numeric/DateTime column from the source table or reducer output. */
  timeColumn: string & keyof TSource
}

export interface ViewOptions<TSource = any> {
  groupBy: GroupByItem | GroupByItem[]
  select: (agg: AggProxy<TSource>) => Record<string, AggExpr | KeyRef>
  /** When set, the MV computes rolling aggregations over the given time window. */
  slidingWindow?: SlidingWindowOptions<TSource>
}

// ─── Type inference from initialState ────────────────────────────

export function inferStateFields(initialState: Record<string, unknown>): StateFieldDef[] {
  const fields: StateFieldDef[] = []
  for (const [name, value] of Object.entries(initialState)) {
    let columnType: string
    let defaultValue: string
    switch (typeof value) {
      case 'number':
        columnType = 'Float64'
        defaultValue = String(value)
        break
      case 'bigint':
        columnType = 'UInt64'
        defaultValue = String(value)
        break
      case 'string':
        columnType = 'String'
        defaultValue = `'${sqlEscape(value as string)}'`
        break
      case 'boolean':
        columnType = 'Boolean'
        defaultValue = value ? 'true' : 'false'
        break
      case 'object':
        columnType = 'Json'
        defaultValue = `'${sqlEscape(JSON.stringify(value))}'`
        break
      default:
        columnType = 'String'
        defaultValue = `'${sqlEscape(String(value))}'`
    }
    fields.push({ name, columnType, defaultValue })
  }
  return fields
}

// ─── DDL generators ──────────────────────────────────────────────

export function tableToSql(
  name: string,
  columns: Record<string, ColumnType>,
  virtual: boolean,
): string {
  const prefix = virtual ? 'CREATE VIRTUAL TABLE' : 'CREATE TABLE'
  const cols = Object.entries(columns)
    .map(([name, ct]) => `${name} ${ct._sql}`)
    .join(', ')
  return `${prefix} ${name} (${cols});`
}

export function reducerToSql(
  name: string,
  source: string,
  groupBy: string[],
  stateFields: StateFieldDef[],
): string {
  const gb = groupBy.join(', ')
  const state = stateFields
    .map((f) => `${f.name} ${f.columnType} DEFAULT ${f.defaultValue}`)
    .join(', ')
  return `CREATE REDUCER ${name} SOURCE ${source} GROUP BY ${gb} STATE (${state}) LANGUAGE EXTERNAL;`
}

export function viewToSql(
  name: string,
  source: string,
  groupByItems: GroupByItem[],
  selectFn: (agg: AggProxy<any>) => Record<string, AggExpr | KeyRef>,
  slidingWindow?: SlidingWindowOptions,
): string {
  const groupByCols: string[] = []
  const intervalDefs: { column: string; seconds: number; alias: string }[] = []

  for (const item of groupByItems) {
    if (typeof item === 'string') {
      groupByCols.push(item)
    } else if (item._type === 'interval') {
      const alias = item.alias || `${item.column}_interval`
      intervalDefs.push({ column: item.column, seconds: item.seconds, alias })
      groupByCols.push(alias)
    }
  }

  const keyProxy = new Proxy({} as Record<string, KeyRef>, {
    get(_target, prop: string): KeyRef {
      return { _type: 'key', column: prop }
    },
  })

  const aggProxy: AggProxy<any> = {
    key: keyProxy,
    sum(col: string) {
      return { _type: 'agg', func: 'sum', column: col }
    },
    count() {
      return { _type: 'agg', func: 'count', column: null }
    },
    first(col: string) {
      return { _type: 'agg', func: 'first', column: col }
    },
    last(col: string) {
      return { _type: 'agg', func: 'last', column: col }
    },
    min(col: string) {
      return { _type: 'agg', func: 'min', column: col }
    },
    max(col: string) {
      return { _type: 'agg', func: 'max', column: col }
    },
    avg(col: string) {
      return { _type: 'agg', func: 'avg', column: col }
    },
  }

  const selectResult = selectFn(aggProxy)
  const selectItems: string[] = []

  for (const [alias, expr] of Object.entries(selectResult)) {
    if (expr._type === 'key') {
      const intv = intervalDefs.find((d) => d.alias === expr.column)
      if (intv) {
        selectItems.push(
          `toStartOfInterval(${intv.column}, INTERVAL ${intv.seconds} SECOND) AS ${alias}`,
        )
      } else {
        selectItems.push(alias === expr.column ? alias : `${expr.column} AS ${alias}`)
      }
    } else if (expr._type === 'agg') {
      const arg = expr.column ? `(${expr.column})` : '()'
      selectItems.push(`${expr.func}${arg} AS ${alias}`)
    }
  }

  let sql = `CREATE MATERIALIZED VIEW ${name} AS SELECT ${selectItems.join(', ')} FROM ${source} GROUP BY ${groupByCols.join(', ')}`

  if (slidingWindow) {
    const seconds = parseDuration(slidingWindow.interval)
    sql += ` WINDOW SLIDING INTERVAL ${seconds} SECOND BY ${slidingWindow.timeColumn}`
  }

  return sql + ';'
}

File: src/pipeline/ddl.rs
//! Builder option types + helpers — Rust mirror of `@settle/stream`'s `ddl.ts`.

use std::collections::HashMap;

use crate::error::{Error, Result};
use crate::types::{Row, RowMap, Value};

// ─── Duration parsing ───────────────────────────────────────────

/// Parse a duration string like `"5 minutes"`, `"1 hour"`, `"86400 seconds"`.
/// Supported units: `s/sec/second(s)`, `m/min/minute(s)`, `h/hr/hour(s)`,
/// `d/day(s)` (case-insensitive).
pub fn parse_duration(s: &str) -> Result<u64> {
    let trimmed = s.trim();
    let split = trimmed
        .find(|c: char| !c.is_ascii_digit())
        .ok_or_else(|| Error::Schema(format!("invalid duration: '{s}'")))?;
    let (num_part, rest) = trimmed.split_at(split);
    let n: u64 = num_part
        .parse()
        .map_err(|_| Error::Schema(format!("invalid duration: '{s}'")))?;
    let unit = rest.trim().to_ascii_lowercase();
    let mult: u64 = match unit.as_str() {
        "s" | "sec" | "second" | "seconds" => 1,
        "m" | "min" | "minute" | "minutes" => 60,
        "h" | "hr" | "hour" | "hours" => 3600,
        "d" | "day" | "days" => 86400,
        _ => return Err(Error::Schema(format!("unknown duration unit: '{unit}'"))),
    };
    Ok(n * mult)
}

// ─── Interval helper ────────────────────────────────────────────

#[derive(Debug, Clone)]
pub struct IntervalExpr {
    pub column: String,
    pub seconds: u64,
    pub alias: Option<String>,
}

impl IntervalExpr {
    pub fn r#as(mut self, alias: impl Into<String>) -> Self {
        self.alias = Some(alias.into());
        self
    }
}

/// `interval("block_time", "5 minutes")` — bucket a DateTime column by duration.
pub fn interval(column: impl Into<String>, duration: &str) -> IntervalExpr {
    let seconds = parse_duration(duration).unwrap_or_else(|e| panic!("interval(): {e}"));
    IntervalExpr {
        column: column.into(),
        seconds,
        alias: None,
    }
}

// ─── Aggregation expressions ────────────────────────────────────

#[derive(Debug, Clone)]
pub struct AggExpr {
    pub func: AggFn,
    pub column: Option<String>,
}

#[derive(Debug, Clone, Copy)]
pub enum AggFn {
    Sum,
    Count,
    First,
    Last,
    Min,
    Max,
    Avg,
}

impl AggFn {
    pub fn name(self) -> &'static str {
        match self {
            AggFn::Sum => "sum",
            AggFn::Count => "count",
            AggFn::First => "first",
            AggFn::Last => "last",
            AggFn::Min => "min",
            AggFn::Max => "max",
            AggFn::Avg => "avg",
        }
    }
}

#[derive(Debug, Clone)]
pub struct KeyRef {
    pub column: String,
}

#[derive(Debug, Clone)]
pub enum Projection {
    Agg(AggExpr),
    Key(KeyRef),
}

impl From<AggExpr> for Projection {
    fn from(a: AggExpr) -> Self {
        Projection::Agg(a)
    }
}
impl From<KeyRef> for Projection {
    fn from(k: KeyRef) -> Self {
        Projection::Key(k)
    }
}

/// Proxy passed to `select: |agg| ...` view callbacks.
pub struct AggProxy;

impl AggProxy {
    pub fn key(&self, column: impl Into<String>) -> KeyRef {
        KeyRef {
            column: column.into(),
        }
    }
    pub fn sum(&self, column: impl Into<String>) -> AggExpr {
        AggExpr {
            func: AggFn::Sum,
            column: Some(column.into()),
        }
    }
    pub fn count(&self) -> AggExpr {
        AggExpr {
            func: AggFn::Count,
            column: None,
        }
    }
    pub fn first(&self, column: impl Into<String>) -> AggExpr {
        AggExpr {
            func: AggFn::First,
            column: Some(column.into()),
        }
    }
    pub fn last(&self, column: impl Into<String>) -> AggExpr {
        AggExpr {
            func: AggFn::Last,
            column: Some(column.into()),
        }
    }
    pub fn min(&self, column: impl Into<String>) -> AggExpr {
        AggExpr {
            func: AggFn::Min,
            column: Some(column.into()),
        }
    }
    pub fn max(&self, column: impl Into<String>) -> AggExpr {
        AggExpr {
            func: AggFn::Max,
            column: Some(column.into()),
        }
    }
    pub fn avg(&self, column: impl Into<String>) -> AggExpr {
        AggExpr {
            func: AggFn::Avg,
            column: Some(column.into()),
        }
    }
}

// ─── GROUP BY items ─────────────────────────────────────────────

#[derive(Debug, Clone)]
pub enum GroupByItem {
    Column(String),
    Interval(IntervalExpr),
}

impl From<&str> for GroupByItem {
    fn from(s: &str) -> Self {
        GroupByItem::Column(s.to_string())
    }
}
impl From<String> for GroupByItem {
    fn from(s: String) -> Self {
        GroupByItem::Column(s)
    }
}
impl From<IntervalExpr> for GroupByItem {
    fn from(i: IntervalExpr) -> Self {
        GroupByItem::Interval(i)
    }
}

// ─── Reducer / view options ─────────────────────────────────────

pub type ReduceFn = Box<dyn Fn(&mut StateCtx, &Row) + Send + Sync>;
pub type SelectFn = Box<dyn FnOnce(&AggProxy) -> Vec<(String, Projection)>>;

pub struct ReducerOptions {
    pub group_by: Vec<String>,
    pub initial_state: RowMap,
    pub reduce: ReduceFn,
}

#[derive(Debug, Clone)]
pub struct SlidingWindowOptions {
    /// Window duration string (`"1 hour"`, `"30 minutes"`, ...).
    pub interval: String,
    /// DateTime/numeric column to window on.
    pub time_column: String,
}

pub struct ViewOptions {
    pub group_by: Vec<GroupByItem>,
    pub sliding_window: Option<SlidingWindowOptions>,
    pub select: SelectFn,
}

// ─── State context (passed to reduce closures) ──────────────────

/// Mutable context handed to `reduce(state, row)` closures.
///
/// Wraps the reducer's running state and accumulates emitted rows for the
/// current input row. Mirrors TS `ReducerCtx<TState, TEmit>` (`state.update()`,
/// `state.emit()`, plus typed read accessors).
pub struct StateCtx<'a> {
    state: &'a mut HashMap<String, Value>,
    emits: Vec<RowMap>,
}

impl<'a> StateCtx<'a> {
    pub(crate) fn new(state: &'a mut HashMap<String, Value>) -> Self {
        Self {
            state,
            emits: Vec::new(),
        }
    }

    pub(crate) fn into_emits(self) -> Vec<RowMap> {
        self.emits
    }

    /// Replace the entire state with the given map.
    pub fn update(&mut self, new_state: RowMap) {
        *self.state = new_state;
    }

    /// Set a single state field without touching the others.
    pub fn set(&mut self, name: impl Into<String>, value: Value) {
        self.state.insert(name.into(), value);
    }

    /// Emit a row into the reducer's output stream.
    pub fn emit(&mut self, row: RowMap) {
        self.emits.push(row);
    }

    /// Read a state field by name.
    pub fn get(&self, name: &str) -> Option<&Value> {
        self.state.get(name)
    }

    pub fn get_f64(&self, name: &str) -> f64 {
        self.state.get(name).and_then(Value::as_f64).unwrap_or(0.0)
    }
    pub fn get_i64(&self, name: &str) -> i64 {
        self.state.get(name).and_then(Value::as_i64).unwrap_or(0)
    }
    pub fn get_u64(&self, name: &str) -> u64 {
        self.state.get(name).and_then(Value::as_u64).unwrap_or(0)
    }
    pub fn get_bool(&self, name: &str) -> bool {
        self.state.get(name).and_then(Value::as_bool).unwrap_or(false)
    }
    pub fn get_str(&self, name: &str) -> &str {
        self.state.get(name).and_then(Value::as_str).unwrap_or("")
    }

    /// Borrow the underlying state map.
    pub fn state(&self) -> &HashMap<String, Value> {
        self.state
    }
}
